save-reporters/src/commonMain/kotlin/org/cqfn/save/reporter/json/JsonReporter.kt

```diff
@@ -0,0 +1,82 @@
+package org.cqfn.save.reporter.json
+
+import org.cqfn.save.core.config.ReportType
+import org.cqfn.save.core.plugin.Plugin
+import org.cqfn.save.core.plugin.PluginException
+import org.cqfn.save.core.reporter.Reporter
+import org.cqfn.save.core.result.TestResult
+
+import okio.BufferedSink
+
+import kotlinx.serialization.encodeToString
+import kotlinx.serialization.json.Json
+
+/**
+ * Reporter that produces a JSON report as a [Report]
+ *
+ * @property out a sink for output
+ */
+class JsonReporter(override val out: BufferedSink) : Reporter {
+    override val type: ReportType = ReportType.JSON
+
+    /**
+     * Formatter to serialize/deserialize to JSON
+     */
+    internal val json = Json
+
+    private var isFirstEvent = true  // todo: use AtomicBoolean
+
+    private var isFirstSuite = true  // todo: use AtomicBoolean
+
+    override fun beforeAll() {
+        out.write("[\n".encodeToByteArray())
+    }
+
+    override fun afterAll() {
+        out.write("]\n".encodeToByteArray())
+    }
+
+    override fun onSuiteStart(suiteName: String) {
+        if (!isFirstSuite) {
+            out.write(",\n".encodeToByteArray())
+        } else {
+            isFirstSuite = false
+        }
```

Comment:
Can we move this part into separate function?
It's duplicates in each method

Reply:
Yeah, tried to refactor it. Looks a little weird, because boolean flag can't be changed from inside the function. When we use atomic, it will look better.